[EXP-38-239] Feat: add gnosis chain support

0xkofee opened the pull request.

0xkofee commented on yearn/db/models.py:
```diff
@@ -35,8 +35,9 @@ class Snapshot(SQLModel, table=True):
+password = os.environ.get('PGPASS', 'yearn')
 
-dsn = f'postgresql://{user}@localhost:5432/yearn'
+dsn = f'postgresql://{user}:{password}@localhost:5432/yearn'
```
this is from another PR I opened

0xkofee commented:
I got `brownie run tvl` to work in this PR

mariuspod reviewed (changes requested):
This is great, could you please add two new exporter service entries into the docker-compose.yml for the dashboards as well (one for forwards and one for historical exports) ?

BobTheBuidler commented:
I just created a PR that will add a checklist for adding new chains. You've done great with this as Marius said, but there are a few more things that will be necessary for the exporters to run. 

Could you take a peek at the checklist here and make sure you've got everything?
https://github.com/yearn/yearn-exporter/pull/242/files

0xkofee commented:
> I just created a PR that will add a checklist for adding new chains. You've done great with this as Marius said, but there are a few more things that will be necessary for the exporters to run.
> 
> Could you take a peek at the checklist here and make sure you've got everything? https://github.com/yearn/yearn-exporter/pull/242/files
- in ./yearn/networks.py:
- [ x ] specify the chainid in the same way as the others.
- [ x ] specify a network label in `Networks.label`.

- in ./yearn/constants.py:
- [ ] specify Treasury address. (if Yearn has a treasury contract deployed) - **No treasury contract deployed**
- [ x ] specify Strategist Multisig address. (if Yearn has a SMS deployed)

- in ./yearn/prices/constants.py:
- [ x ] add weth address to `tokens_by_network`.
- [ x ] add usdc address to `tokens_by_network`.
- [ x ] add dai address to `tokens_by_network`. - **NOTE: Main currency on gnosis is already Dai(xDai). Do we need this? Currently set to 0x44fA8E6f47987339850636F88629646662444217**
- [ x ] configure `stablecoins` using popular + "safe" stablecoins on the chain.

- in ./yearn/multicall2.py:
- [ x ] specify the appropriate multicall2 contract address. - **NOTE: Deployment not from maker. Maker contract not verified. Diff looks good**

- in ./yearn/prices/chainlink.py:
- [ x ] specify any chainlink feeds available on the chain in `feeds`.
- [ ] specify the chainlink registry address in `registries` if applicable, or specify None if not applicable. - **Note: No registry found **

- in ./yearn/prices/uniswap/v2.py:
- [ x ] specify the factory and router address for any important uni v2 forks on the chain.

- in ./yearn/utils.py:
- [ x ] specify `BINARY_SEARCH_BARRIER` as 0 if all historical data is available. If historical data not available due to a network upgrade/etc, specify the first block of the available historical data.

- in ./yearn/middleware/middleware.py:
- [ x ] specify `BATCH_SIZE` of approx 1 day based on avg block times on the chain.

- in ./scripts/historical-exporter.py:
- [ x ] configure `end` to equal the date + hour of deployment for the first Yearn product on the chain. **NOTE: I assume yearn vault based off this post: https://github.com/yearn/yearn-exporter/issues/227#issuecomment-1029510526**
- [ x ] configure `data_query`. See `mapping` in ./yearn/outputs/victoria/output_helper.py to figure out the metric to specify based on what type of product was the first deployed on the chain.

- in ./scripts/historical-treasury-exporter.py:
- [ ] configure `end` to equal the date + hour of deployment for the Yearn Treasury on the chain.
- [ ] configure `data_query` using the existing networks as an example.

- in ./scripts/historical-sms-exporter.py:
- [ ] configure `end` to equal the date + hour of deployment for the Strategist Multisig on the chain. **NOTE: No treasury deployed**
- [ ] configure `data_query` using the existing networks as an example.


You also need to set up containers for each exporter on the new chain. This is a little more complicated but you can use existing containers as an example:
- [ ] define common envs for the chain
- [ ] add one new service entry for forwards exporter
- [ ] add one new service entry for historical exporter
- [ ] add one new service entry for treasury exporter
- [ ] add one new service entry for historical treasury exporter
- [ ] add one new service entry for sms exporter
- [ ] add one new service entry for historical sms exporter
- [ ] add one new service entry for wallet exporter
- [ ] add one new service entry for transactions exporter
- [ ] adapt entrypoint.sh

BobTheBuidler commented:
> * NOTE: Main currency on gnosis is already Dai(xDai). Do we need this? Currently set to 0x44fA8E6f47987339850636F88629646662444217

Good question! For Gnosis chain it should be fine to use wxDAI (0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d) rather than DAI as it will be more widely used on the chain. Please include both in the stablecoins constant.

> * configure `end` to equal the date + hour of deployment for the first Yearn product on the chain. **NOTE: I assume yearn vault based off this post: **

In this case, you're correct that the first product deployed is a yearn vault. But that is not always the case, on Fantom for example, the first Yearn product deployed was an IB market. (Yes, I know IB is now separate but that is neither here nor there...) What you did looks just fine. 

Without a Yearn Treasury live on Gnosis chain, do you know which address is set to receive perf/mgmt fees from the Gnosis vault deployment?

0xkofee commented:
> > * NOTE: Main currency on gnosis is already Dai(xDai). Do we need this? Currently set to 0x44fA8E6f47987339850636F88629646662444217
> 
> Good question! For Gnosis chain it should be fine to use wxDAI (0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d) rather than DAI as it will be more widely used on the chain. Please include both in the stablecoins constant.
> 
> > * configure `end` to equal the date + hour of deployment for the first Yearn product on the chain. **NOTE: I assume yearn vault based off this post: **
> 
> In this case, you're correct that the first product deployed is a yearn vault. But that is not always the case, on Fantom for example, the first Yearn product deployed was an IB market. (Yes, I know IB is now separate but that is neither here nor there...) What you did looks just fine.
> 
> Without a Yearn Treasury live on Gnosis chain, do you know which address is set to receive perf/mgmt fees from the Gnosis vault deployment?

19. governance →
[0x22eae41c7da367b9a15e942eb6227df849bb498c](https://blockscout.com/xdai/mainnet/address/0x22eae41c7da367b9a15e942eb6227df849bb498c)
20. management →
[0xfb4464a18d18f3ff439680bbbce659db2806a187](https://blockscout.com/xdai/mainnet/address/0xfb4464a18d18f3ff439680bbbce659db2806a187)
21. guardian →
[0x5fcdc32dfc361a32e9d5ab9a384b890c62d0b8ac](https://blockscout.com/xdai/mainnet/address/0x5fcdc32dfc361a32e9d5ab9a384b890c62d0b8ac)

0xkofee commented:
> This is great, could you please add two new exporter service entries into the docker-compose.yml for the dashboards as well (one for forwards and one for historical exports) ?

by forwards do you mean the exporter.py?

BobTheBuidler commented:
> > This is great, could you please add two new exporter service entries into the docker-compose.yml for the dashboards as well (one for forwards and one for historical exports) ?
> 
> by forwards do you mean the exporter.py?

Sorry for the delay, just got back from Denver. 

Yes you are correct, forwards exporter would be exporter.py

0xkofee commented:
Hi @BobTheBuidler I believe I addressed most of the points. I left notes on stuff I couldn't address. I am able to run the following commands successfully
`brownie run tvl`
`brownie run print_strategies`

brownie config:
```
xdai-archive
    ├─id: xdai-archive
    ├─chainid: 100
    ├─explorer: https://blockscout.com/xdai/mainnet/api
    └─host: https://xdai-archive.blockscout.com
```

BobTheBuidler commented:
@0xkofee I added a few more things to your branch, and am working on testing now. Do you know of a good public archive node for Gnosis chain?

0xkofee commented:
@BobTheBuidler Hey there! I messaged you on telegram but wasn't sure if you had PM's turned off. anyways I was using this archive node: 
```
xdai-archive
    ├─id: xdai-archive
    ├─chainid: 100
    ├─explorer: https://blockscout.com/xdai/mainnet/api
    └─host: https://xdai-archive.blockscout.com
```

BobTheBuidler commented:
Closing as this is replaced by #289
